### ISSUE TITLE
Implement state management in Nextflow channels

### ISSUE THREAD

**Issue by rcannood** [viash-io/viash#479]
We currently have multiple approaches to process / split / join channels in our Nextflow workflows.

## Approach 0

Do everything manually with `map` and `flatMap` and `filter`

## Approach 1: `setWorkflowArguments` - `getWorkflowArguments`

Using these helper functions, we can specify where argument values have to go a priori.

Benefit: 

* Can reuse a category in multiple components, if need be

Downsides:

* No way of defining how component outputs gets stored or mapped to the next consumer in the workflow. After a component is run, it typically returns a single file instead of a map. The `getWorkflowArguments()` thus needs to infer what the name of the upcoming argument will be and renames it to `input`.
* Constantly moving data from the second element and third element in the tuple.
* Nesting workflows requires recursively growing the tuple.

Example ([source](https://github.com/openpipelines-bio/openpipeline/blob/c5ed662a8916ef2d2c1a63e01166a11a9f113c36/workflows/multiomics/integration/initialize_integration/main.nf)):

```nextflow
workflow run_wf {
  take:
  input_ch

  main:
  output_ch = input_ch
    | preprocessInputs("config": config)
    // FORMAT: [
    //   id, 
    //   [input: ..., obsm_pca: ..., ...]
    // ]
    
    // split params for downstream components
    | setWorkflowArguments(
      pca: [
        "input": "input", 
        "obsm_output": "obsm_pca",
        "var_input": "var_pca_feature_selection",
        "modality": "modality"
      ],
      neighbors: [
        "uns_output": "uns_neighbors",
        "obsp_distances": "obsp_neighbor_distances",
        "obsp_connectivities": "obsp_neighbor_connectivities",
        "modality": "modality"
      ],
      umap: [ 
        "uns_neighbors": "uns_neighbors",
        "output": "output",
        "obsm_output": "obsm_umap",
        "modality": "modality"
      ]
    )
    // FORMAT: [
    //   id,
    //   [],
    //   [pca: [input: ...], neighbors: [...], umap: [...]]
    // ]
    
    | getWorkflowArguments(key: "pca")
    // FORMAT: [
    //   id,
    //   [input: ..., obsm_output: ..., var_input: ..., modality: ...],
    //   [pca: [...], neighbors: [...], umap: [...]]
    // ]

    | pca
    // FORMAT: [
    //   id,
    //   pca_output,
    //   [pca: [...], neighbors: [...], umap: [...]]
    // ]

    | getWorkflowArguments(key: "neighbors")
    // FORMAT: [
    //   id,
    //   [input: ..., uns_output: ..., obsp_distances: ..., obsp_connectivities: ..., modality: ...],
    //   [pca: [...], neighbors: [...], umap: [...]]
    // ]

    | find_neighbors
    | getWorkflowArguments(key: "umap")

    | umap.run(
      auto: [ publish: true ],
      args: [ output_compression: "gzip" ]
    )
    // FORMAT: [
    //   id,
    //   umap_output,
    //   [pca: [...], neighbors: [...], umap: [...]]
    // ]

    // remove splitArgs
    | map { tup ->
      tup.take(2) + tup.drop(3)
    }
    // FORMAT: [
    //   id,
    //   umap_output,
    // ]

  emit:
  output_ch
}
```

**Comment by rcannood** [comment]
## Proposed approach: `from_state` - `to_state`

Channel only contains an `[id, state]` event. Each component will draw data from the state and publish new data to the state.

Benefits:

* Since the channel can only ever contain tuples with just an id and the state, it's easier to understand because you don't need as much documentation on how what the format of the tuple is at any point in time
* Less function calls

Downsides:

* Need to manually specify what happens to the output of each component

Example:

```nextflow
workflow run_wf {
  take:
  input_ch

  main:
  output_ch = input_ch
    | preprocessInputs("config": config)
    // FORMAT: [
    //   id, 
    //   [input: ..., obsm_pca: ..., ...]
    // ]
    
    | pca.run(
      from_state: [
        input: "input",
        obsm_output: "obsm_pca",
        var_input: "var_pca_feature_selection",
        modality: "modality"
      ],
      to_state: [
        data: "output"
      ]
    )

    | find_neighbors.run(
      from_state: [
        input: "data",
        uns_output: "uns_neighbors",
        obsp_distances: "obsp_neighbor_distances",
        obsp_connectivities: "obsp_neighbor_connectivities",
        modality: "modality"
      ],
      to_state: [
        data: "output"
      ]
    )

    | umap.run(
      from_state: { id, state -> [
        input: state.data,
        uns_neighbors: state.uns_neighbors,
        output: state.output,
        obsm_output: state.obsm_umap,
        modality: state.modality,
        output_compression: "gzip"
      ]},
      to_state: [
        data: "output"
      ],
      auto: [publish: true]
    )
```

The `from_state` argument can be any of the following formats:

* List of strings (e.g. ["input", "foo", "bar"]`): The component will take the elements with those names from the state and passes it to the underlying workflow
* Map (e.g. `[input: "file", foo: "bar"]`): The component takes the item values from the state, renames them to the key values, and passes them to the underlying workflow
* Closure (e.g. `{id, state -> [input: state.file, foo: state.bar] }`): Custom processing of the id and state using a closure.

The `to_state` argument can also be a List of strings, a Map, or a Closure.

* Should the `to_state` always be additive or not? We could have an extra argument `additive` to determine whether the `to_state` is additive or not. Example:
  - `umap.run(to_state: ["data"], auto: [additive: true])` adds a field called `output` to the current state
  - `umap.run(to_state: ["data"], auto: [additive: false])` replaces the state with a map `[data: ...]`.

## Split states

We could just use a `flatMap`? Do we gain anything by creating custom helper functions?

```nextflow
  | flatMap{ id, state ->
    ["rna", "prot", "atac"].collect{ key ->
      def new_id = id + "_" + key
      def new_state = [input: state[key], orig_id: id]   
      [new_id, new_state]
    }
  )
```

## Join states

Add helper functions for joining states. This operation can be M-1 or M-N.

### Example `M-1`

```nextflow
  | group_by{ id, state -> "combined" }
  | map_states{ id, states ->
       def new_state = [
        input: states.collect{it.data}
        output: states[0].output
      ]
      new_state
    }
  )
```

### Example `M-N`

```nextflow
  | group_by{ id, state -> state.some_string }
  | map_states{ id, states ->
      def new_state = [
        input: states.collect{it.data}
        output: states[0].output
      ]
      new_state
    }
  )
```

Input: 

```yaml
- ["one", [input: "one.h5ad", some_string: "foo", output: "foo.h5ad"]]
- ["two", [input: "two.h5ad", some_string: "foo", output: "foo.h5ad"]]
- ["three", [input: "three.h5ad", some_string: "bar", output: "bar.h5ad"]]
```

Output:

```yaml
- ["foo", [input: ["one.h5ad", "two.h5ad"], output: "foo.h5ad"]]
- ["bar", [input: ["three.h5ad"], output: "bar.h5ad"]]
```

## Publish states

It might make sense to add a helper function to save the state in its entirety. For example:

```nextflow
  | publish_states()
```

Which creates the following files:

* `$params.publish_dir/$id/state.yaml`: A yaml of the current state, containing:
  ```yaml
  input: myinput.h5mu
  data: myoutput.h5mu
  obsm_pca: pca
  var_pca_feature_selection: hvg
  ...
  ```

* `$params.publish_dir/$id/myinput.h5mu`: The input file (because it was in the state)
* `$params.publish_dir/$id/myoutput.h5mu`: The output file (because it was in the state)

You could add arguments to this function to allow choosing the output path and format of the state json / yaml, and possibly remove keys. E.g.

```nextflow
  | publish_states(
    state_file: "state.json",
    state_file_format: "json",
    exclude: ["input"]
  )
```

## Multiplex

How to run multiple components in parallel?

We could implement a helper function `run_components`.

Example:

```nextflow
    | run_components(
      components: [umap, tsne, mds],
      from_state: {id, state, config ->
        ...
      },
      to_state: {id, data, config ->
        [
          component_id: config.functionality.name,
          data: data.output
        ]
      },
    )
```

The `from_state` and `to_state` arguments function similar to the similarly named arguments in `component.run`, except that the closure takes a third argument, which will contain the viash config of the component.

**Comment by rcannood** [comment]
Closed by #501, will be included in Viash 0.7.5.
